src/coreclr/jit/lclvars.cpp

```diff
@@ -1709,6 +1702,26 @@ bool Compiler::StructPromotionHelper::CanPromoteStructType(CORINFO_CLASS_HANDLE
     unsigned structAlignment = roundUp(compHandle->getClassAlignmentRequirement(typeHnd), TARGET_POINTER_SIZE);
 #endif // TARGET_ARM
 
+    // If we have "Custom Layout" then we might have an explicit Size attribute
+    // Managed C++ uses this for its structs, such C++ types will not contain GC pointers.
+    //
+    // The current VM implementation also incorrectly sets the CORINFO_FLG_CUSTOMLAYOUT
+    // whenever a managed value class contains any GC pointers.
+    // (See the comment for VMFLAG_NOT_TIGHTLY_PACKED in class.h)
+    //
+    // It is important to struct promote managed value classes that have GC pointers
+    // So we compute the correct value for "CustomLayout" here
+    //
+    if (StructHasCustomLayout(typeFlags) && ((typeFlags & CORINFO_FLG_CONTAINS_GC_PTR) == 0))
```

Comment:
I am not sure if we can include `CORINFO_FLG_BYREF_LIKE` here. It goes to this code https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/tools/Common/Internal/Runtime/MethodTable.Constants.cs#L171
 @MichalStrehovsky do you know if we can promote this classes as a single byref?

Reply:
Both TypedReference and Span have multiple fields - I don't know how much struct promotion helps for that. There's a C# feature in the works to support ref fields; the runtime already supports it. It allows for `ref struct MyStruct { ref int RefForInt; }`. Maybe it would be beneficial for that, but I don't know how common that pattern will be.

Reply:
I see, thanks